nautobot/extras/tests/test_webhooks.py

```diff
@@ -261,3 +268,54 @@ class FakeResponse:
 
     def test_webhook_render_body_with_utf8(self):
         self.assertEqual(Webhook().render_body({"utf8": "I am UTF-8! 😀"}), '{"utf8": "I am UTF-8! 😀"}')
+
+    @patch("nautobot.extras.tasks.process_webhook.apply_async")
+    def test_enqueue_webhooks(self, mock_async):
+        request_id = uuid.uuid4()
+        self.client.force_login(self.user)
+
+        with web_request_context(self.user, change_id=request_id):
+            location_type = LocationType.objects.get(name="Campus")
+            location = Location(name="Location 1", location_type=location_type, status=self.statuses[0])
+            location.save()
+
+            mock_async.assert_called_once()
+            args = mock_async.call_args[1]["args"]
+            self.assertEqual(args[0], Webhook.objects.get(type_create=True).pk)
+            self.assertEqual(args[1]["name"], "Location 1")
+            self.assertEqual(args[2], "location")
+            self.assertEqual(args[3], ObjectChangeActionChoices.ACTION_CREATE)
+            self.assertEqual(args[5], self.user.username)
+            self.assertEqual(args[6], request_id)
+            self.assertEqual(args[7]["prechange"], None)
+            self.assertEqual(args[7]["postchange"]["name"], "Location 1")
+            self.assertEqual(args[7]["differences"]["removed"], None)
+            self.assertEqual(args[7]["differences"]["added"]["name"], "Location 1")
+
+    @patch("nautobot.extras.tasks.process_webhook.apply_async")
+    def test_enqueue_webhooks_m2m_update_with_pruned_object_changes(self, mock_async):
+        # https://github.com/nautobot/nautobot/issues/4327
+        request_id = uuid.uuid4()
+        self.client.force_login(self.user)
+        location_type = LocationType.objects.get(name="Campus")
+        location = Location(name="Location 1", location_type=location_type, status=self.statuses[0])
+        location.save()
+
+        tag = Tag.objects.create(name="Tag 1")
+
+        all_changes = get_changes_for_model(location)
+        # Mimicking when all changes have been pruned via CHANGELOG_RETENTION
+        all_changes.delete()
+
+        with web_request_context(self.user, change_id=request_id):
+            location.tags.add(tag)
```

Comment:
Doesn't/shouldn't this result in a new ObjectChange being created and therefore `most_recent_change` (below) being non-None? Not sure I fully understand the code flow involved here.

Reply:
Maybe the bug here is that modifying M2M fields on an object does not trigger an actual change log to be created, but it **does** trigger the webhook:
```
In [1]: Location.objects.first()
Out[1]: <Location: Australia>

In [2]: location = _

In [3]: location.tags.all()
Out[3]: <TagQuerySet []>

In [4]: Tag.objects.first()
Out[4]: <Tag: Amber>

In [5]: tag = _

In [6]: from nautobot.core.utils.lookup import get_changes_for_model

In [7]: changes = get_changes_for_model(location)

In [8]: changes.delete()
Out[8]: (1, {'extras.ObjectChange': 1})

In [9]: from nautobot.extras.context_managers import web_request_context

In [10]: user = User.objects.get(username="admin")

In [11]: with web_request_context(user, context_detail="manual-fix"):
    ...:     location.tags.add(tag)
    ...:

In [12]: get_changes_for_model(location)
Out[12]: <RestrictedQuerySet []>
```

Reply:
My fix is merely the bandaid.

Reply:
That *sort of* makes sense, as via the UI or API you're only going to change `tags` or any other M2M via an operation that also saves the object itself (even if nothing about the object has otherwise changed). Once #3321 is fixed though that would probably expose this issue anew. 

This might be possible to fix even now, I suspect the following code in `nautobot/extras/signals.py` to be the issue:

```python
        if object_m2m_changed:
            related_changes = ObjectChange.objects.filter(
                changed_object_type=ContentType.objects.get_for_model(instance),
                changed_object_id=instance.pk,
                request_id=change_context_state.get().change_id,
            )
            m2m_changes = instance.to_objectchange(action)
            related_changes.update(object_data=m2m_changes.object_data, object_data_v2=m2m_changes.object_data_v2)   # <-----<< will do nothing if there's no existing objectchange
            objectchange = related_changes.first() if related_changes.exists() else None  # <-----<< else case should probably create a new objectchange instead of using None.
```

I also note the following inconsistent behavior in that same function:

```python
        # Enqueue job hooks
        if objectchange is not None:
            enqueue_job_hooks(objectchange)

    # Enqueue webhooks
    enqueue_webhooks(instance, change_context_state.get().get_user(), change_context_state.get().change_id, action)
```

I suspect you'll find that calling `object.tags.set(...)` when there's no existing objectchange will trigger webhooks but will not trigger jobhooks. 😿 

Would you like to take a stab at addressing either or both of the above issues as a part of this PR?